### PR TITLE
Safer handling of Accept headers

### PR DESCRIPTION
**Pull request by mibanescu**
closes #3303
https://pulp.plan.io/issues/3303

### CODE REVIEW

File: crane/views/v2.py [hunk 1]
```diff
@@ -77,7 +78,8 @@ def name_redirect(relative_path):
         if schema2_data or manifest_list_data:
             # if it is a newer docker client it sets accept headers to manifest schema 1, 2 and list
             # if it is an older docker client, he doesnot set any of accept headers
-            accept_headers = request.headers.get('Accept')
+            accept_headers = get_accept_headers(request)
+            log.debug("Accept: %s", accept_headers)
```
Comment (ipanova) [inline crane/views/v2.py]: please remove the log statement, this information will not help anyone without a deep understanding of docker engine insides

File: crane/views/v2.py [hunk 2]
```diff
@@ -125,3 +127,12 @@ def handle_error(error):
     response.headers['Content-Type'] = 'application/json'
     response.status_code = error.status_code
     return response
+
+
+def get_accept_headers(request):
+    accept_headers = request.headers.get('Accept')
+    if not accept_headers:
+        return []
+    log.debug("Accept: %s", accept_headers)
```
Comment (ipanova) [inline crane/views/v2.py]: same here

File: crane/views/v2.py [hunk 3]
```diff
@@ -125,3 +127,12 @@ def handle_error(error):
...
+    accept_headers = accept_headers.split(',')
+    return set(x.strip() for x in accept_headers)
```
Comment (ipanova) [inline crane/views/v2.py]: for header in accept_header i find easier to read, in general i would avoid x,y,i, etc variable names.

File: crane/views/v2.py [hunk 4]
```diff
@@ -125,3 +127,12 @@ def handle_error(error):
     response.headers['Content-Type'] = 'application/json'
     response.status_code = error.status_code
     return response
+
+
+def get_accept_headers(request):
```
Comment (ipanova) [inline crane/views/v2.py]: please add doc block

File: tests/views/test_path.py [hunk 5]
```diff
@@ -14,12 +14,15 @@ def test_invalid_repo_name(self):
         self.assertEqual(parsed_response_data['errors'][0]['message'], 'Not Found')
 
     def test_valid_repo_name_for_manifest(self):
-        headers = {'Accept': 'application/vnd.docker.distribution.manifest.v2+json'}
-        response = self.test_client.get('/v2/redhat/foo/manifests/1.25.1-musl', headers=headers)
+        # #3303: verify multi-valued headers too
+        # manifest lists are evaluated first, so pass a longer media type that
```
Comment (ipanova) [inline tests/views/test_path.py]: can you elaborate more this please?

File: tests/views/test_path.py [hunk 6]
```diff
@@ -14,12 +14,15 @@ def test_invalid_repo_name(self):
...
+        # matches the manifest list as a prefix
+        headers = {'Accept': 'application/vnd.docker.distribution.manifest.list.v2+jsonjunk,application/vnd.docker.distribution.manifest.v2+json'}  # noqa
```
Comment (ipanova) [inline tests/views/test_path.py]: i don't think +jsonjunk would be a valid mediatype sent from docker client :) max what you can have is +json or +prettyjws

File: crane/views/v2.py (hunk 1 above)
Comment (bmbouter) [reply crane/views/v2.py]: I don't anything about docker/crane, but I think someone that someone who turns on debug logging would have that deep understanding and want logging like this.

File: tests/views/test_path.py (hunk 5 above)
Comment (mibanescu) [reply tests/views/test_path.py]: The code first checks if manifest_list_mediatype is in accept_headers. If the client sent a string that *includes* 'application/vnd.docker.distribution.manifest.list.v2+json like in my +junk test, as well as a valid manifest v2 media type, then you'd get a false match for a manifest list.

File: crane/views/v2.py (hunk 3 above)
Comment (mibanescu) [reply crane/views/v2.py]: I would agree, but for comprehension I find it concise enough. If you care deeply I can change it.

File: crane/views/v2.py (hunk 1 above)
Comment (mibanescu) [reply crane/views/v2.py]: I agree with bmbouter. More than once I wanted to have more information, and I would prefer to do that without changing code (just changing config). That being said, I shouldn't log.debug twice.

Comment (ipanova) [reply crane/views/v2.py]: it is not about docker/crane but about docker-engine, the client for pulling the docker content and docker api. I still kinda think this is not very helpful, but if we want to keep this then please at least update to a more meaningful message something like " Accept headers sent from docker client" or "Accept header sent during docker pull"

File: tests/views/test_path.py (hunk 6 above)
Comment (mibanescu) [reply tests/views/test_path.py]: That's with a civilized client. As a server, you should always check the inputs from the client. This is not a realistic example, just like this patch is not fixing something broken *now*.

File: crane/views/v2.py (hunk 3 above)
Comment (ipanova) [reply crane/views/v2.py]: not really. it's just my personal preference of the pattern.

File: crane/views/v2.py (hunk 2 above)
Comment (ipanova) [reply crane/views/v2.py]: what about leaving just one log statement with "Received Accept headers" or similar?

File: crane/views/v2.py (hunk 4 above)
Comment (mibanescu) [reply crane/views/v2.py]: Done.

**Comment by mibanescu** [comment]
All requested changes have been done.

**Pull request merged**